src/ipfs.js

```diff
@@ -0,0 +1,136 @@
+const execFile = require('child_process').execFile
+const { getIPNS } = require('./ethereum')
+const { isHashBlacklisted, isParcelBlacklisted } = require('./blacklist')
+const { setParcel, getParcel, setIPFS, getIPFS } = require('./database')
+const request = require('request')
+
+module.exports = class Download {
+  constructor () {
+    this.download = async (req, res) => {
+      try {
+        const ipfs = req.params.ipfs
+        const file = req.params[0] ? `${ipfs}/${req.params[0]}` : ipfs
+
+        const blackListed = await isHashBlacklisted(ipfs)
+        if (blackListed) {
+          throw new Error(`IPFS ${ipfs} is blacklisted`)
+        }
+
+        request.get(`http://localhost:8080/ipfs/${file}`).pipe(res)
+      } catch (error) {
+        return res.json({ ok: false, error: error.message })
+      }
+    }
+    this.pin = async (req, res) => {
+      try {
+        const [x, y] = [req.params.x, req.params.y]
+        const ipns = await getIPNS(x, y)
+        if (!ipns) {
+          throw new Error('The land has not IPNS associated')
+        }
+        await Download.connectToPeer(req.params.peerId)
+        const ipfs = await Download.resolveIPNS(ipns)
+        await Download.publishHash(ipfs)
+        const dependencies = await Download.resolveDependencies(ipfs)
+        await setIPFS(ipns, ipfs)
+        await setParcel({ x, y }, { ipns, ipfs, dependencies })
+        return res.json({ ok: true, message: 'Pinning Success' })
+      } catch (error) {
+        return res.json({ ok: false, error: error.message })
+      }
+    }
+    this.resolve = async (req, res) => {
+      try {
+        const [x, y] = [req.params.x, req.params.y]
+        const blackListed = await isParcelBlacklisted(x, y)
+        if (blackListed) {
+          throw new Error(`Parcel (${x},${y}) is blacklisted`)
+        }
+
+        const cachedResponse = await getParcel(x, y)
+        if (cachedResponse && !req.query.force) {
+          return res.json({ ok: true, url: cachedResponse })
+        }
+
+        const ipns = await getIPNS(x, y)
+        if (!ipns) {
+          throw new Error('The land has not IPNS associated')
+        }
+        const ipfs = await Download.resolveIPNS(ipns)
+        const dependencies = await Download.resolveDependencies(ipfs)
+        const url = { ipns, ipfs, dependencies }
+        await setParcel({ x, y }, url)
+        return res.json({ ok: true, url })
+      } catch (error) {
+        return res.json({ ok: false, error: error.message })
```

Comment:
does it add a status code `403`? I cannot recall this.

Reply:
Yeah, lets add an errorHandling too